src/index.ts

```diff
@@ -619,6 +620,7 @@ export async function assemble(config: Configuration, components: Components): P
   // Create `Program`
   const program = {
     ...shorthands,
+    ...{ ...Events.listenTo(fsEvents), ...Events.listenTo(sessionEvents)},
```

Comment:
Doesn't this override the file system events with the session events?
I think it might be better to just create one overarching type for the events and listen to that?
```ts
// events.ts
type All = Events.FileSystem & Events.Session<Session>

// index.ts
const events = Events.createEmitter<Events.All>()
...Events.listenTo(events)
```
Also means you have to pass less instances around which is nice.

That said, I'm not sure if you can use a subset as a param with this 🤔 
For example, if you create a Session and pass in the "all" emitter,
would you be able to use the following as the type of the constructor param?
```ts
Events.Emitter<Events.Session<Session>>
```
I would think that would work, because it is a subset of the `All` type and you only use the session events. Thoughts?

Reply:
> Doesn't this override the file system events with the session events?

Yes, it does. Good catch. 🙏 

> I think it might be better to just create one overarching type for the events and listen to that?
> 
> ```ts
> // events.ts
> type All = Events.FileSystem & Events.Session<Session>
> 
> // index.ts
> const events = Events.createEmitter<Events.All>()
> ...Events.listenTo(events)
> ```

We can't use the `Session` type in `events.ts` because that would create a circular dependency. What we could do is:

> 
> ```ts
> // events.ts
> export type All<S> = FileSystem & Session<S>
> 
> // index.ts
> const events = Events.createEmitter<Events.All<Session>>()
> ...Events.listenTo(events)
> ```

---

> That said, I'm not sure if you can use a subset as a param with this 🤔 For example, if you create a Session and pass in the "all" emitter, would you be able to use the following as the type of the constructor param?
> 
> ```ts
> Events.Emitter<Events.Session<Session>>
> ```
> 
> I would think that would work, because it is a subset of the `All` type and you only use the session events. Thoughts?

Unfortunately, this does not appear to work out. Not entirely sure why, perhaps the rules are different within a generic?

One option is to pass around `all` events everywhere, but we'd be losing some type information on which events different parts of the code emit. I suspect that we will have to pass `all` around anyways because that is what will be listened upon. But perhaps we can do something with `extends` or something to document which parts of `all` are used. Or maybe there is a way to construct `all` from `fileSystem` and `session` emitters somehow. 🤔

Reply:
What if you keep the individual emitters, but then write a function to merge them together to use with `Events.listen()`?
```ts
merge(a: EventEmitter<A>, b: EventEmitter<B>): EventEmitter<A & B> 
```
Just realised I repeated what you said 🙈

Reply:
Started experimenting with this a bit, but it seems like it pushes the typing issues into the merge function. Here's what I have so far:

```typescript
export function merge<A, B>(a: EventEmitter<A>, b: EventEmitter<B>): EventEmitter<A & B>  {
  const merged = createEmitter<A & B>()
  const aEmit = a.emit
  const bEmit = b.emit

  a.emit = (...args) => {
    aEmit.call(a, ...args)
    merged.emit(...args)
  }

  b.emit = (...args) => {
    bEmit.call(b, ...args)
    merged.emit(...args)
  }

  return merged
}
```

The idea is to rewire each event emitter's `emit` function so that it also calls `merged.emit`. But we end up with type errors on `merged.emit` like:

```
(parameter) args: [eventName: K, event: A[K]]
Argument of type 'A[K]' is not assignable to parameter of type '(A & B)[K]'.
  Type 'A' is not assignable to type 'A & B'.
    Type 'A' is not assignable to type 'B'.
      'B' could be instantiated with an arbitrary type which could be unrelated to 'A'.
```

Reply:
```ts
  a.emit = <K extends keyof A>(eventName: K, event: (A & B)[ K ]) => {
    aEmit.call(a, eventName, event)
    merged.emit(eventName, event)
  }

  b.emit = <K extends keyof B>(eventName: K, event: (A & B)[ K ]) => {
    bEmit.call(b, eventName, event)
    merged.emit(eventName, event)
  }
  ```
  This seems to work?

Reply:
Yesss! That did it! 🙌 

Pushed the changes up. Thanks for the help thinking through this. 🙏